File: HandSpecifications/TwoPairSpecification.php
<?php

class TwoPairSpecification extends HandSpecification {

    /**
     * @param Hand $Hand
     * @return boolean
     */
    public function isSatisfiedBy(Hand $Hand) {
        $numPairsSeen = 0;
        $CardsGroupedByValues = $Hand->getCardsGroupedByValues();
        foreach ($CardsGroupedByValues as $Cards) {
            if (count($Cards) == 2) {
                $numPairsSeen++;
            }
        }

        $canMakeTwoPair = (2 == $numPairsSeen);
        return $canMakeTwoPair;
    }
}

File: HandSpecifications/TwoPairSpecification.cs
class TwoPairSpecification : HandSpecification {

    public override bool isSatisfiedBy(Hand Hand) {
        int numPairsSeen = 0;
        var CardsGroupedByValues = Hand.getCardsGroupedByValues();
        foreach (var Cards in CardsGroupedByValues) {
            if (Cards.Count == 2) {
                numPairsSeen++;
            }
        }

        var canMakeTwoPair = (2 == numPairsSeen);
        return canMakeTwoPair;
    }

    public override Hand newHand(Hand Hand) {
        return new TwoPair(Hand.getCards());
    }
}